wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabOAuthTokenFetcher.java

```diff
@@ -48,33 +54,53 @@
       ImmutableSet.of("api", "write_repository", "openid");
   public static final String OAUTH_2_PREFIX = "oauth2-";
 
+  private final List<String> registeredGitlabEndpoints;
   private final OAuthAPI oAuthAPI;
   private final String apiEndpoint;
-  private final GitlabApiClient gitlabApiClient;
 
   @Inject
   public GitlabOAuthTokenFetcher(
       @Nullable @Named("che.integration.gitlab.server_endpoints") String gitlabEndpoints,
+      @Nullable @Named("che.integration.gitlab.oauth_endpoint") String oauthEndpoint,
       @Named("che.api") String apiEndpoint,
       OAuthAPI oAuthAPI) {
     this.apiEndpoint = apiEndpoint;
-    this.oAuthAPI = oAuthAPI;
     if (gitlabEndpoints != null) {
-      final String oAuthEndpoint = Splitter.on(",").splitToList(gitlabEndpoints).get(0);
-      this.gitlabApiClient = new GitlabApiClient(StringUtils.trimEnd(oAuthEndpoint, '/'));
+      this.registeredGitlabEndpoints =
+          Splitter.on(",")
+              .splitToStream(gitlabEndpoints)
+              .map(e -> StringUtils.trimEnd(e, '/'))
+              .collect(toList());
+    } else {
+      this.registeredGitlabEndpoints = Collections.emptyList();
+    }
+    if (oauthEndpoint != null) {
+      if (!registeredGitlabEndpoints.contains(StringUtils.trimEnd(oauthEndpoint, '/'))) {
+        throw new ConfigurationException(
+            "GitLab OAuth integration endpoint must be present in registered GitLab endpoints list.");
+      }
+      this.oAuthAPI = oAuthAPI;
     } else {
-      this.gitlabApiClient = null;
+      this.oAuthAPI = null;
     }
   }
 
   @Override
   public PersonalAccessToken fetchPersonalAccessToken(Subject cheSubject, String scmServerUrl)
       throws ScmUnauthorizedException, ScmCommunicationException {
     scmServerUrl = StringUtils.trimEnd(scmServerUrl, '/');
+    GitlabApiClient gitlabApiClient = getApiClient(scmServerUrl);
```

Comment:
why don't we made this as filed?

Reply:
we need dedicated client for each registered url. so keeping them all in memory is wasteful. imo better to create on-demand.